File: wasm/verify/tabview_test.mjs
// The TabView bar and page selection, in the runtime.
//
// core.TabView's four wire props (tabs, selectedIndex, onTabChange, one child
// per page) went entirely unread here until this pass: a TabView was a bare box
// holding every page at once, with no bar and no way to switch, while both
// natives drew a bar above the selected page alone. These tests hold the
// runtime to the behaviour htmlout/tabview.go states for the pair, plus the two
// things only the live target can get wrong — the click dispatch, and the
// bookkeeping that keeps a bar from throwing the child indices out of step.

import test from "node:test";
import assert from "node:assert/strict";

import { loadRuntime, nodeAt } from "./load.mjs";

const page = (text) => ({ Type: "Column", Children: [{ Type: "Text", Props: { content: text } }] });

const TABS = [
    { label: "Home", icon: "🏠" },
    { label: "Search", icon: "🔍" },
];

// A TabView at root, with as many pages as it has tabs.
function mountTabs({ selectedIndex = 0, onTabChange = "cb_tab", tabs = TABS } = {}) {
    const rt = loadRuntime();
    const props = { selectedIndex, tabs };
    if (onTabChange) props.onTabChange = onTabChange;
    rt.GrMob.mount(
        JSON.stringify({
            Type: "TabView",
            Props: props,
            Children: tabs.map((t) => page(`${t.label} page`)),
        })
    );
    rt.drainFrames();
    const root = rt.mountPoint.children[0];
    return {
        rt,
        root,
        bar: () => root.children[0],
        // The pages, by *node* index — which is what every path and every
        // patch means, and is one more than the DOM index while a bar exists.
        pageAt: (i) => nodeAt(rt.document, `root/${i}`),
    };
}

// --------------------------------------------------------------------------
// The bar
// --------------------------------------------------------------------------

test("a TabView draws a bar ahead of its pages", () => {
    const { root, bar } = mountTabs();

    assert.equal(root.children.length, 3, "bar plus two pages");
    assert.equal(bar().dataset.grmobChrome, "tabbar");
    assert.equal(bar().getAttribute("role"), "tablist");
    // Chrome is not a node: nothing may be addressed to it.
    assert.equal(bar().getAttribute("data-node-path"), null);
    assert.equal(bar().children.length, 2);
    assert.deepEqual(
        bar().children.map((b) => b.textContent),
        ["Home", "Search"]
    );
    assert.deepEqual(
        bar().children.map((b) => b.getAttribute("role")),
        ["tab", "tab"]
    );
});

// The icon half of a core.TabItem is drawn by no target: Compose's Tab is
// built with `text = { Text(label) }` and the SwiftUI bar with a Text of the
// label. Drawing it here would make the web the outlier instead of closing a
// gap.
test("a tab shows its label and not its icon", () => {
    const { bar } = mountTabs();
    for (const button of bar().children) {
        assert.ok(!button.textContent.includes("🏠"), `icon leaked into ${button.textContent}`);
    }
});

test("a TabView with no tabs draws no bar at all", () => {
    const { root } = mountTabs({ tabs: [] });
    assert.equal(root.children.length, 0);
});

test("the selected tab is the one marked selected", () => {
    const { bar } = mountTabs({ selectedIndex: 1 });
    assert.deepEqual(
        bar().children.map((b) => b.getAttribute("aria-selected")),
        ["false", "true"]
    );
    assert.equal(bar().children[1].style.fontWeight, "600");
    assert.equal(bar().children[0].style.fontWeight, "");
});

// The callback ID is mirrored onto the bar so the live DOM carries the same
// data-ontabchange htmlout writes into an export. The listener is what
// dispatches; this is the record of it.
test("the bar records the callback ID, and drops it when there is none", () => {
    assert.equal(mountTabs().bar().getAttribute("data-ontabchange"), "cb_tab");
    assert.equal(mountTabs({ onTabChange: null }).bar().getAttribute("data-ontabchange"), null);
});

// --------------------------------------------------------------------------
// Selection
// --------------------------------------------------------------------------

test("only the selected page is visible", () => {
    const { pageAt } = mountTabs({ selectedIndex: 1 });
    assert.equal(pageAt(0).style.display, "none");
    // Not "" — a Column is a flex container whether or not its Style asks
    // (stackAxisFor), and a page restored by clearing the declaration would
    // come back in block flow.
    assert.equal(pageAt(1).style.display, "flex");
});

// Both natives guard the page rather than clamping the index
// (`children.indices.contains` in Swift, `getOrNull` in Kotlin), and the Swift
// bar compares `i == selected` with no clamp either.
test("an out-of-range selection shows no page and marks no tab", () => {
    const { bar, pageAt } = mountTabs({ selectedIndex: 7 });
    assert.equal(pageAt(0).style.display, "none");
    assert.equal(pageAt(1).style.display, "none");
    assert.deepEqual(
        bar().children.map((b) => b.getAttribute("aria-selected")),
        ["false", "false"]
    );
});

test("a click on a tab dispatches its index to onTabChange", () => {
    const { rt, bar } = mountTabs();

    bar().children[1].dispatch("click");

    // A number, not a string: ReceiveEventPayload sniffs the value's type, and
    // only a number reaches the int callback map core.OnTabChange registered in.
    assert.deepEqual(rt.dispatched, [{ id: "cb_tab", payload: { value: 1 } }]);
});

test("a bar whose callback was pruned goes inert rather than dispatching a stale ID", () => {
    const { rt, root, bar } = mountTabs();

    // The pass stopped carrying onTabChange. An update-props patch sends the
    // whole new map, so its absence is definitive — pruneStaleListeners drops
    // the ID, and callback IDs are positional, so keeping it would fire some
    // other node's handler.
    rt.GrMob.patch(
        JSON.stringify([
            { Type: "update-props", TargetID: "root", Changes: { selectedIndex: 0, tabs: TABS } },
        ])
    );
    bar().children[1].dispatch("click");

    assert.deepEqual(rt.dispatched, []);
    assert.equal(root.children[0].getAttribute("data-ontabchange"), null);
});

// --------------------------------------------------------------------------
// Switching tabs
// --------------------------------------------------------------------------

// core.SelectedIndex is controlled state, so a switch reaches the page as a
// prop patch on the TabView and never as a subtree replacement: the pages are
// never rebuilt, only re-hidden.
test("a selectedIndex patch moves both the indicator and the visible page", () => {
    const { rt, bar, pageAt } = mountTabs({ selectedIndex: 0 });

    rt.GrMob.patch(
        JSON.stringify([
            {
                Type: "update-props",
                TargetID: "root",
                Changes: { selectedIndex: 1, tabs: TABS, onTabChange: "cb_tab" },
            },
        ])
    );

    assert.equal(pageAt(0).style.display, "none");
    assert.equal(pageAt(1).style.display, "flex");
    assert.deepEqual(
        bar().children.map((b) => b.getAttribute("aria-selected")),
        ["false", "true"]
    );
});

// The bar is rebuilt only when the tab strip itself changed. A switch is
// exactly when a keyboard user has one of these buttons focused, and a rebuild
// would throw that focus away on every tab press.
test("switching tabs keeps the same button elements", () => {
    const { rt, bar } = mountTabs({ selectedIndex: 0 });
    const before = bar().children[0];

    rt.GrMob.patch(
        JSON.stringify([
            {
                Type: "update-props",
                TargetID: "root",
                Changes: { selectedIndex: 1, tabs: TABS, onTabChange: "cb_tab" },
            },
        ])
    );

    assert.equal(bar().children[0], before, "the bar was rebuilt on a mere selection change");
});

test("a changed tab strip does rebuild the bar", () => {
    const { rt, bar } = mountTabs();
    const renamed = [{ label: "Home" }, { label: "Explore" }, { label: "Me" }];

    rt.GrMob.patch(
        JSON.stringify([
            {
                Type: "update-props",
                TargetID: "root",
                Changes: { selectedIndex: 0, tabs: renamed, onTabChange: "cb_tab" },
            },
        ])
    );

    assert.deepEqual(
        bar().children.map((b) => b.textContent),
        ["Home", "Explore", "Me"]
    );
    // Still ahead of the pages, so the chrome offset holds.
    assert.equal(bar().dataset.grmobChrome, "tabbar");
});

// --------------------------------------------------------------------------
// The hiding against the rest of the runtime
// --------------------------------------------------------------------------

// styleFromGrMob is *total*: it assigns every property it manages on every
// pass, so an update-style on a hidden page rewrites its display. The
// selection is recomputed at the end of each batch for exactly this reason.
test("an update-style on a hidden page does not reveal it", () => {
    const { rt, pageAt } = mountTabs({ selectedIndex: 0 });

    rt.GrMob.patch(
        JSON.stringify([
            { Type: "update-style", TargetID: "root/1", Changes: { Background: "#eee" } },
        ])
    );

    assert.equal(pageAt(1).style.display, "none", "a style patch unhid a page");
    assert.equal(pageAt(1).style.background, "#eee", "the style patch itself was lost");
});

// The visible page's display comes back from what the style pass computed, not
// from clearing the declaration.
test("a style patch on the visible page leaves it visible", () => {
    const { rt, pageAt } = mountTabs({ selectedIndex: 0 });

    rt.GrMob.patch(
        JSON.stringify([
            { Type: "update-style", TargetID: "root/0", Changes: { Padding: { Top: 8, Right: 8, Bottom: 8, Left: 8 } } },
        ])
    );

    assert.equal(pageAt(0).style.display, "flex");
});

// --------------------------------------------------------------------------
// The chrome offset
// --------------------------------------------------------------------------
//
// Patch paths address a child by its *node* index, and the bar makes the DOM
// index one higher. Both places that convert have to shift past the chrome, or
// a page lands in front of the bar and every path after it is out of step.

test("an added page lands after the bar, at the slot its path names", () => {
    const { rt, root, bar } = mountTabs({ selectedIndex: 0 });

    rt.GrMob.patch(
        JSON.stringify([
            { Type: "add", TargetID: "root/0", Changes: page("inserted") },
        ])
    );

    assert.equal(root.children[0], bar(), "the bar was pushed out of first place");
    assert.equal(root.children[1].getAttribute("data-node-path"), "root/0");
    assert.equal(nodeAt(rt.document, "root/0"), root.children[1]);
    assert.equal(root.children[1].children[0].textContent, "inserted");
});

test("add-child numbers the new page from the pages, not from the DOM", () => {
    const { rt, root } = mountTabs({ selectedIndex: 0 });

    rt.GrMob.patch(
        JSON.stringify([
            { Type: "add-child", TargetID: "root", Changes: page("third") },
        ])
    );

    // Two pages existed, so the new one is page 2 — not 3, which is what the
    // DOM child count alone would have said.
    assert.equal(root.children.length, 4);
    assert.equal(root.children[3].getAttribute("data-node-path"), "root/2");
    assert.equal(nodeAt(rt.document, "root/2"), root.children[3]);
});

// A TabView with no bar must not be given a phantom offset — the offset is
// counted from the children, not assumed from the node type.
test("a TabView with no tabs numbers its pages from zero", () => {
    const rt = loadRuntime();
    rt.GrMob.mount(
        JSON.stringify({ Type: "TabView", Props: { selectedIndex: 0 }, Children: [page("only")] })
    );
    rt.drainFrames();
    const root = rt.mountPoint.children[0];

    rt.GrMob.patch(
        JSON.stringify([{ Type: "add-child", TargetID: "root", Changes: page("second") }])
    );

    assert.equal(root.children[1].getAttribute("data-node-path"), "root/1");
});

// --------------------------------------------------------------------------
// The tab/panel wiring
// --------------------------------------------------------------------------
//
// A well-formed tablist still says nothing about which region each tab
// governs. That relationship needs ids, and the ids are derived from the node
// path so they are the same strings htmlout writes — TestTabsAndPanelsPointAtEachOther
// in htmlout/tabview_test.go asserts these exact literals for the same tree.

test("a tab and its page point at each other", () => {
    const { root, bar, pageAt } = mountTabs();

    for (let i = 0; i < 2; i++) {
        const tab = bar().children[i];
        assert.equal(tab.getAttribute("id"), `grmob-root-tab-${i}`);
        assert.equal(tab.getAttribute("aria-controls"), `grmob-root-panel-${i}`);

        const page = pageAt(i);
        assert.equal(page.getAttribute("id"), `grmob-root-panel-${i}`);
        assert.equal(page.getAttribute("role"), "tabpanel");
        assert.equal(page.getAttribute("aria-labelledby"), `grmob-root-tab-${i}`);
    }
    // Every reference resolves inside the tree, which is the property the ids
    // exist for: a dangling IDREF reads as a region that is not there.
    for (const tab of bar().children) {
        const target = root.findAll((el) => el.getAttribute("id") === tab.getAttribute("aria-controls"));
        assert.equal(target.length, 1, `aria-controls of ${tab.textContent} names no single element`);
    }
});

// Ids are document-global. The scope is the node path, so a nested TabView
// cannot collide with the one above it — and does not need a counter to avoid
// it, which is what keeps these ids equal to the exporter's.
test("a nested TabView gets its own id scope", () => {
    const rt = loadRuntime();
    rt.GrMob.mount(
        JSON.stringify({
            Type: "Column",
            Children: [
                { Type: "TabView", Props: { selectedIndex: 0, tabs: TABS }, Children: [page("a"), page("b")] },
                { Type: "TabView", Props: { selectedIndex: 0, tabs: TABS }, Children: [page("c"), page("d")] },
            ],
        })
    );
    rt.drainFrames();
    const root = rt.mountPoint.children[0];

    const ids = root.findAll((el) => el.getAttribute("id") !== null).map((el) => el.getAttribute("id"));
    assert.equal(new Set(ids).size, ids.length, `duplicate ids: ${ids}`);
    assert.ok(ids.includes("grmob-root-0-panel-0"), `scope is not the node path: ${ids}`);
    assert.ok(ids.includes("grmob-root-1-panel-0"), `scope is not the node path: ${ids}`);
});

// role="tabpanel" would replace the role the browser already gives a <button>,
// an <img> or an <input>, so those pages are left alone — and the tab must not
// point at one, since aria-controls naming a non-panel is a lie about the
// document. GENERIC_TAGS is the set, pinned to Go's by
// TestRuntimeGenericTagsMatchGo.
test("a page whose element already has a role is not made a panel", () => {
    const rt = loadRuntime();
    rt.GrMob.mount(
        JSON.stringify({
            Type: "TabView",
            Props: { selectedIndex: 0, tabs: TABS },
            Children: [page("home"), { Type: "Button", Props: { label: "press" } }],
        })
    );
    rt.drainFrames();
    const root = rt.mountPoint.children[0];
    const bar = root.children[0];

    assert.equal(root.children[2].getAttribute("role"), null, "a <button> page was given the panel role");
    assert.equal(root.children[2].getAttribute("id"), null);
    assert.equal(bar.children[1].getAttribute("aria-controls"), null, "its tab still claims to control it");
    // The eligible page beside it keeps its wiring: one page opting out is not
    // the tab set opting out.
    assert.equal(bar.children[0].getAttribute("aria-controls"), "grmob-root-panel-0");
});

// The author took the page out of the accessibility tree on purpose. Read off
// the element rather than out of the Style, because that is where
// applyAccessibility put it and this runs long after.
test("an aria-hidden page is not made a panel", () => {
    const { rt, root, pageAt } = mountTabs();

    rt.GrMob.patch(
        JSON.stringify([
            { Type: "update-style", TargetID: "root/1", Changes: { AccessibilityHidden: true } },
        ])
    );

    assert.equal(pageAt(1).getAttribute("aria-hidden"), "true");
    assert.equal(pageAt(1).getAttribute("role"), null);
    assert.equal(root.children[0].children[1].getAttribute("aria-controls"), null);
});

// The wiring is derived state like the selection, so it has to survive the
// same three ways a batch can invalidate it. Here: a page becomes eligible
// again after having been ruled out, which a guarded write would never undo.
test("a page that becomes eligible again is rewired", () => {
    const { rt, root, pageAt } = mountTabs();

    const hide = (v) =>
        rt.GrMob.patch(
            JSON.stringify([
                { Type: "update-style", TargetID: "root/1", Changes: { AccessibilityHidden: v } },
            ])
        );
    hide(true);
    assert.equal(pageAt(1).getAttribute("role"), null);

    hide(false);
    assert.equal(pageAt(1).getAttribute("role"), "tabpanel");
    assert.equal(pageAt(1).getAttribute("id"), "grmob-root-panel-1");
    assert.equal(root.children[0].children[1].getAttribute("aria-controls"), "grmob-root-panel-1");
});

// aria-labelledby wins over aria-label in the accessible-name calculation, so
// writing it unconditionally would discard the name the app author chose. The
// tab still points at the panel; only the naming is left alone.
test("a page that names itself keeps its name", () => {
    const { rt, root, pageAt } = mountTabs();

    rt.GrMob.patch(
        JSON.stringify([
            { Type: "update-style", TargetID: "root/0", Changes: { AccessibilityLabel: "Your home feed" } },
        ])
    );

    assert.equal(pageAt(0).getAttribute("aria-label"), "Your home feed");
    assert.equal(pageAt(0).getAttribute("aria-labelledby"), null, "the tab's name overrode the author's");
    assert.equal(pageAt(0).getAttribute("role"), "tabpanel", "the page stopped being a panel because it had a name");
    assert.equal(root.children[0].children[0].getAttribute("aria-controls"), "grmob-root-panel-0");
});

// A page no tab names is not part of a tab set. The count comes from the bar,
// so a tabs prop that shrinks takes the surplus page's wiring with it.
test("a page with no tab of its own is not a panel", () => {
    const { rt, root, pageAt } = mountTabs();

    rt.GrMob.patch(
        JSON.stringify([
            {
                Type: "update-props",
                TargetID: "root",
                Changes: { selectedIndex: 0, tabs: [{ label: "Home" }], onTabChange: "cb_tab" },
            },
        ])
    );

    assert.equal(root.children[0].children.length, 1, "the bar was not rebuilt from the new strip");
    assert.equal(pageAt(0).getAttribute("role"), "tabpanel");
    assert.equal(pageAt(1).getAttribute("role"), null, "a page no tab names was left wired");
    assert.equal(pageAt(1).getAttribute("id"), null);
});

// No bar, no tab set, nothing to wire — and no ids to collide with a real
// TabView's elsewhere in the document.
test("a TabView with no tabs wires no panels", () => {
    const rt = loadRuntime();
    rt.GrMob.mount(
        JSON.stringify({ Type: "TabView", Props: { selectedIndex: 0 }, Children: [page("only")] })
    );
    rt.drainFrames();
    const root = rt.mountPoint.children[0];

    assert.equal(root.children[0].getAttribute("role"), null);
    assert.equal(root.children[0].getAttribute("id"), null);
});

// The hidden pages are wired too: the relationship is a property of the
// document's structure, not of the current selection, so a switch moves the
// indicator and the visibility and leaves the wiring exactly where it was.
test("a switch leaves the wiring alone", () => {
    const { rt, pageAt } = mountTabs({ selectedIndex: 0 });

    rt.GrMob.patch(
        JSON.stringify([
            {
                Type: "update-props",
                TargetID: "root",
                Changes: { selectedIndex: 1, tabs: TABS, onTabChange: "cb_tab" },
            },
        ])
    );

    for (let i = 0; i < 2; i++) {
        assert.equal(pageAt(i).getAttribute("role"), "tabpanel");
        assert.equal(pageAt(i).getAttribute("aria-labelledby"), `grmob-root-tab-${i}`);
    }
});

// A page added by a patch arrives already wired, because the wiring rides on
// the same end-of-batch pass the selection does.
test("a page added by a patch is wired like the rest", () => {
    const { rt, root } = mountTabs();

    rt.GrMob.patch(
        JSON.stringify([
            {
                Type: "update-props",
                TargetID: "root",
                Changes: {
                    selectedIndex: 0,
                    tabs: [...TABS, { label: "Me" }],
                    onTabChange: "cb_tab",
                },
            },
            { Type: "add-child", TargetID: "root", Changes: page("third") },
        ])
    );

    assert.equal(root.children[3].getAttribute("role"), "tabpanel");
    assert.equal(root.children[3].getAttribute("id"), "grmob-root-panel-2");
    assert.equal(root.children[0].children[2].getAttribute("aria-controls"), "grmob-root-panel-2");
});

File: htmlout/tabview_test.go
package htmlout

import (
	"strings"
	"testing"

	"github.com/rohanthewiz/grmob/core"
)

// A TabView with two tabs and one page each, the shape core.TabView builds.
func tabViewNode(selected int, onChange string) *core.Node {
	props := map[string]any{
		"selectedIndex": selected,
		"tabs": []map[string]string{
			{"label": "Home", "icon": "🏠"},
			{"label": "Search", "icon": "🔍"},
		},
	}
	if onChange != "" {
		props["onTabChange"] = onChange
	}
	return &core.Node{
		Type:  "TabView",
		Props: props,
		Children: []*core.Node{
			{Type: "Column", Children: []*core.Node{{Type: "Text", Props: map[string]any{"content": "home page"}}}},
			{Type: "Column", Children: []*core.Node{{Type: "Text", Props: map[string]any{"content": "search page"}}}},
		},
	}
}

// The gap this closed: none of core.TabView's four wire props were read here,
// so a TabView exported as a bare box holding every page, with no bar — while
// both natives drew a bar above the selected page alone.
func TestTabViewExportsABar(t *testing.T) {
	out := ExportHTML(tabViewNode(0, "cb_9"))
	for _, want := range []string{
		`data-grmob-chrome="tabbar"`,
		`role="tablist"`,
		`data-ontabchange="cb_9"`,
		`role="tab"`,
		`data-tab-index="0"`,
		`data-tab-index="1"`,
		">Home<",
		">Search<",
	} {
		if !strings.Contains(out, want) {
			t.Errorf("missing %q:\n%s", want, out)
		}
	}
}

// The bar comes first. It is the order the runtime's chrome offset assumes,
// and the order a screen reader and a style-less rendering want anyway.
func TestTabBarPrecedesThePages(t *testing.T) {
	out := ExportHTML(tabViewNode(0, "cb_9"))
	bar := strings.Index(out, `data-grmob-chrome="tabbar"`)
	page := strings.Index(out, "home page")
	if bar < 0 || page < 0 {
		t.Fatalf("want both a bar and a page:\n%s", out)
	}
	if bar > page {
		t.Errorf("the bar was written after the first page:\n%s", out)
	}
}

// The icon is drawn by no target: Compose's Tab is built with
// `text = { Text(label) }` and the SwiftUI bar with a Text of the label, so
// drawing it here would make the web the outlier rather than close a gap.
func TestTabBarDrawsLabelsAndNotIcons(t *testing.T) {
	out := ExportHTML(tabViewNode(0, "cb_9"))
	if strings.Contains(out, "🏠") || strings.Contains(out, "🔍") {
		t.Errorf("an icon reached the export:\n%s", out)
	}
}

func TestSelectedTabIsMarkedAndStyled(t *testing.T) {
	out := ExportHTML(tabViewNode(1, "cb_9"))
	tabs := tabTags(t, out)
	if len(tabs) != 2 {
		t.Fatalf("want two tabs, got %d:\n%s", len(tabs), out)
	}
	if !strings.Contains(tabs[0], `aria-selected="false"`) {
		t.Errorf("tab 0 is not marked unselected:\n%s", out)
	}
	if strings.Contains(tabs[0], tabSelectedChassis) {
		t.Errorf("tab 0 carries the selected styling:\n%s", out)
	}
	if !strings.Contains(tabs[1], `aria-selected="true"`) {
		t.Errorf("tab 1 is not marked selected:\n%s", out)
	}
	if !strings.Contains(tabs[1], tabSelectedChassis) {
		t.Errorf("the selected tab carries no selected styling:\n%s", out)
	}
}

// tabTags returns each tab button's opening tag, in document order. Sliced
// rather than matched with a regexp because everything asserted about a tab
// lives in one tag and the boundary is unambiguous: element writes the whole
// tag on one line, and the label follows the ">".
func tabTags(t *testing.T, out string) []string {
	t.Helper()
	var tags []string
	rest := out
	for {
		at := strings.Index(rest, `<button type="button" role="tab"`)
		if at < 0 {
			return tags
		}
		rest = rest[at:]
		end := strings.Index(rest, ">")
		if end < 0 {
			t.Fatalf("unterminated tab tag:\n%s", out)
		}
		tags = append(tags, rest[:end])
		rest = rest[end:]
	}
}

// The pages are hidden, not dropped. The runtime cannot drop one — its patch
// TargetIDs are positional, so its DOM has to stay isomorphic to the node tree
// — and an export that silently lost every screen but one would be a worse
// document for no gain.
func TestOnlyTheSelectedPageIsVisible(t *testing.T) {
	out := ExportHTML(tabViewNode(1, ""))
	if !strings.Contains(out, "home page") || !strings.Contains(out, "search page") {
		t.Fatalf("a page was dropped rather than hidden:\n%s", out)
	}
	home := strings.Index(out, "home page")
	search := strings.Index(out, "search page")
	// The hidden page's own box carries the declaration; find the style
	// attribute immediately preceding its text.
	if !strings.Contains(out[:home], tabHiddenPage) {
		t.Errorf("the unselected page was not hidden:\n%s", out)
	}
	if strings.Contains(out[home:search], tabHiddenPage) {
		t.Errorf("the selected page was hidden:\n%s", out)
	}
}

// display:none has to outrank the display:flex a stack container is given
// unconditionally (see stackAxes), which is what the last-one-wins ordering of
// the declaration list buys.
func TestHiddenPageKeepsItsOwnStyleAndLoses(t *testing.T) {
	out := ExportHTML(tabViewNode(1, ""))
	home := strings.Index(out, "home page")
	decls := out[:home]
	flex := strings.LastIndex(decls, "display:flex")
	none := strings.LastIndex(decls, "display:none")
	if flex < 0 || none < 0 {
		t.Fatalf("want both declarations on the hidden page:\n%s", out)
	}
	if none < flex {
		t.Errorf("display:none precedes display:flex, so the page is not hidden:\n%s", out)
	}
}

// Both natives guard the page rather than clamping the index
// (`children.indices.contains` in Swift, `getOrNull` in Kotlin), and the Swift
// bar compares `i == selected` with no clamp either.
func TestOutOfRangeSelectionShowsNoPage(t *testing.T) {
	out := ExportHTML(tabViewNode(7, ""))
	if strings.Count(out, tabHiddenPage) != 2 {
		t.Errorf("want both pages hidden:\n%s", out)
	}
	if strings.Contains(out, `aria-selected="true"`) {
		t.Errorf("a tab was marked selected for an out-of-range index:\n%s", out)
	}
}

// A bar with no tabs to name would be a hairline across the top of the node,
// which is a visible artifact of an absent prop.
func TestNoTabsMeansNoBar(t *testing.T) {
	n := &core.Node{Type: "TabView", Children: []*core.Node{{Type: "Text", Props: map[string]any{"content": "only"}}}}
	out := ExportHTML(n)
	if strings.Contains(out, "tabbar") {
		t.Errorf("a bar was drawn for a TabView with no tabs:\n%s", out)
	}
	// And the one page is page 0, so it stays visible.
	if strings.Contains(out, tabHiddenPage) {
		t.Errorf("the only page was hidden:\n%s", out)
	}
}

// A TabView with no onTabChange — core.TabView omits the prop for a nil
// handler, keeping a static strip diff-stable — records no callback.
func TestNoHandlerMeansNoCallbackAttribute(t *testing.T) {
	if out := ExportHTML(tabViewNode(0, "")); strings.Contains(out, "data-ontabchange") {
		t.Errorf("a callback attribute was written for a handler-less TabView:\n%s", out)
	}
}

// A tab page with no box of its own still has to be hidden: renderNode forwards
// the parent's declaration through the transparent branch to the children
// standing in for it.
func TestATransparentPageIsStillHidden(t *testing.T) {
	n := tabViewNode(0, "")
	n.Children[1] = &core.Node{Type: "Fragment", Children: []*core.Node{
		{Type: "Text", Props: map[string]any{"content": "fragment page"}},
	}}
	out := ExportHTML(n)
	at := strings.Index(out, "fragment page")
	if at < 0 {
		t.Fatalf("the fragment page was dropped:\n%s", out)
	}
	// The <span> holding it is what carries the hiding, since the Fragment
	// itself has no element.
	span := strings.LastIndex(out[:at], "<span")
	if span < 0 || !strings.Contains(out[span:at], tabHiddenPage) {
		t.Errorf("a transparent page's children were not hidden:\n%s", out)
	}
}

// The tabs prop after a JSON round trip, which is what a tree handed to this
// exporter from outside Go looks like.
func TestTabsSurviveAJSONShapedPropMap(t *testing.T) {
	n := &core.Node{
		Type: "TabView",
		Props: map[string]any{
			"selectedIndex": float64(1),
			"tabs": []any{
				map[string]any{"label": "One"},
				map[string]any{"label": "Two"},
			},
		},
		Children: []*core.Node{
			{Type: "Text", Props: map[string]any{"content": "first"}},
			{Type: "Text", Props: map[string]any{"content": "second"}},
		},
	}
	out := ExportHTML(n)
	for _, want := range []string{">One<", ">Two<"} {
		if !strings.Contains(out, want) {
			t.Errorf("missing %q:\n%s", want, out)
		}
	}
	first := strings.Index(out, "first")
	if !strings.Contains(out[:first], tabHiddenPage) {
		t.Errorf("the float64 selectedIndex was not read:\n%s", out)
	}
}

// The TabView box itself keeps everything any other container gets: its stack
// default, its own Style, and its accessibility attributes. The bar is added to
// that, not substituted for it.
func TestTabViewKeepsItsOwnBox(t *testing.T) {
	n := tabViewNode(0, "")
	n.Style = &core.Style{Background: "#fff"}
	out := ExportHTML(n)
	for _, want := range []string{"background:#fff", "display:flex", "flex-direction:column"} {
		if !strings.Contains(out, want) {
			t.Errorf("missing %q on the TabView box:\n%s", want, out)
		}
	}
}

// Spacer takes an early return, before the shared attribute assembly, so it is
// the one node type that could silently ignore what its parent imposed on it.
// A Spacer page is an odd thing to build, but "every page but one is hidden"
// has to be true of every page.
func TestASpacerPageIsStillHidden(t *testing.T) {
	n := tabViewNode(0, "")
	n.Children[1] = &core.Node{Type: "Spacer", Props: map[string]any{"size": 40}}
	out := ExportHTML(n)
	at := strings.Index(out, "height:40px")
	if at < 0 {
		t.Fatalf("the Spacer page was dropped:\n%s", out)
	}
	// The whole of the tag the size landed in, so the check reads that one
	// element's declaration list and not a neighbour's.
	open := strings.LastIndex(out[:at], "<div")
	close := strings.Index(out[at:], ">")
	if open < 0 || close < 0 {
		t.Fatalf("unterminated Spacer tag:\n%s", out)
	}
	if !strings.Contains(out[open:at+close], tabHiddenPage) {
		t.Errorf("the Spacer page was not hidden:\n%s", out)
	}
}
